Autocast support for scaled_dot_product_attention

mikekgfb opened the pull request:
Summary: Autocast support for scaled_dot_product_attention

Test Plan: sandcastle and guthub cicd

Differential Revision: D42085525



cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

ngimel commented on aten/src/ATen/autocast_mode.cpp:
```diff
@@ -389,6 +389,8 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
   KERNEL(rnn_tanh_cell, lower_precision_fp)
   KERNEL(rnn_relu_cell, lower_precision_fp)
+  KERNEL(_scaled_dot_product_flash_attention, lower_precision_fp)
+  KERNEL(_scaled_dot_product_attention, lower_precision_fp)
```
I think we need only `_scaled_dot_product_flash_attention`, also we need a dedicated test (because previously this failure wasn't caught).

mikekgfb replied:
I updated the PR to include tests that test that flash is failing when run with FP32 and works with autocast to float16 and bfloat16

mikekgfb commented:
@pytorchbot merge

drisspg commented on test/test_transformers.py:
```diff
@@ -1292,6 +1292,19 @@ def test_unaligned_tensors(self):
+    def test_flash_fail_fp32t(self):
+        device = 'cuda'
+        dtype = torch.float
+        size = (2, 2, 8, 5)
```
Flash doesn't accept head dim size of 5 so I don't this is failing for dtype 32 which we do already test with. If you decorate with auto cast does this enable flash?

mikekgfb commented:
@pytorchbot merge

drisspg reviewed (approved):
👍

Commit:
Autocast support for scaled_dot_product_attention (#91066)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/91066

Autocast support for scaled_dot_product_attention

Test Plan: sandcastle and guthub cicd

Reviewed By: drisspg

Differential Revision: D42085525

fbshipit-source-id: 58adc0daf573bdb628f040babb123e8100c7fbd3

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge -f "ignore cancelled check"
